0000-friends_in_high_kindednesses.md

```diff
@@ -0,0 +1,478 @@
+more complex kinds, such as `type -> type`. An example of a term of this kind
+is `Vec`, which takes a type as a parameter and evaluates to a type. The
+difference between the kind of `Vec` and the kind of `Vec<i32>` (which is
+`type`) is analogous to the difference between the type of `not` and
+`not(true)`. Note that `Vec<T>` has the kind `type`, just like `Vec<i32>`: even
+though `T` is a type parameter, `Vec` is still being applied to a type, just
+like `not(x)` still has the type `bool` even though `x` is a variable.
+
+A relatively uncommon feature of Rust is that it has _two_ base kinds, whereas
+many languages which deal with higher-kindedness only have the base kind
+`type`. The other base kind of Rust is the lifetime parameter. If you have a
+type like `Foo<'a>`, the kind of `Foo` is `lifetime -> type`.
+
+Higher-kinded terms can take multiple arguments as well, of course. `Result`
+has the kind `type, type -> type`. Given `vec::Iter<'a, T>` `vec::Iter` has the
+kind `lifetime, type -> type`.
+
+Terms of a higher kind are often called 'type operators'; the type operators
+which evaluate to a type are called 'type constructors'. There are other type
+operators which evaluate to other type operators, and there are even higher
+order type operators, which take type operators as their argument (so they have
+a kind like `(type -> type) -> type`). This RFC doesn't deal with anything as
+exotic as that.
+
+Specifically, the goal of this RFC is to allow type constructors to be
+associated with traits, just as you can currently associate functions, types,
+and consts with traits. There are other forms of polymorphism involving type
+constructors, such as implementing traits for a type constructor instead of a
+type, which are not a part of this RFC.
+
+## Features of associated type constructors
+
+### Declaring an associated type constructor
+
+This RFC proposes a very simple syntax for defining an associated type
+constructor, which looks a lot like the syntax for creating aliases for type
+constructors. The goal of using this syntax is to avoid to creating roadblocks
+for users who do not already understand higher kindedness.
+
+```rust
+trait StreamingIterator {
+   type Item<'a>;
+}
+```
+
+Here, it is clear that `Item` is a type constructor, because it carries a
+parameter. Associated type constructors can carry any number of type and
+lifetime parameters, as in:
+
+```rust
+trait FooBar {
+    type Baz<'a, T, U>;
+}
+```
+
+Associated type constructors can be followed by `where` clauses, which place
+trait bounds on the types constructed by this constructor. For example:
+
+```rust
+trait Collection<T> {
+    type Iter<'a> where for<'a> Self::Iter<'a>: Iterator<Item=&'a T>;
+    type IterMut<'a> where for<'a> Self::IterMut<'a>: Iterator<Item=&'a mut T>;
+    type IntoIter: Iterator<Item=T>;
+}
+```
+
+A `where` clause is used to avoid the impression that this is providing a
```

Comment:
This raises some questions which are not answered in this section of the RFC. On functions, `where` clauses make that function optional when the `where` clause is not met. That is not the case here. Is it a good idea to use internal `where` clauses with majorly different semantics?

Are these `where` clauses anything but sugar over adding to the `where` clause at the head of the trait? Are they restricted to HRTBs for this type constructor, or could they contain arbitrary additional clauses?

Is there a more fitting syntax for this feature which meets all of these conditions?
* The clause is clearly and easily connected to the associated type constructor (e.g. by being a part of the associated type constructor's declaration)
* The clause is unambiguously distinguished from higher-kinded trait bounds on the associated type constructor itself (for forward compatibility).
* The clause is self-evidently restricted to providing HRTB bounds on the types produced by this associated type constructor.